### ISSUE TITLE
How to export posec3d to ONNX?

### ISSUE THREAD

**Issue by vsantosu** [open-mmlab/mmaction2#955]
Hello there,

First of all thanks for the great work! It is deeply appreciated.

I'm trying to export posec3d to ONNX but there I come across the following error:

```
RuntimeError: Expected 5-dimensional input for 5-dimensional weight [32, 17, 1, 7, 7], but got 4-dimensional input of size [544, 1, 7, 7] instead
```

The script command looks like:

```
python tools/pytorch2onnx.py configs/skeleton/posec3d/slowonly_r50_u48_240e_ntu60_xsub_keypoint.py \
                            checkpoints/slowonly_r50_u48_240e_ntu60_xsub_keypoint.pth \
                            --shape 32 17 1 7 7 \
                            --verify
```

Any idea what is wrong?

Thanks in advance!

**Comment by kennymckormick** [comment]
Currently PoseC3D does not support ONNX, we will consider adding this feature in the future.

**Comment by vsantosu** [comment]
Thanks for the quick reply!

So it is not supported because the ONNX operation set does not support some layers or it is just not implemented to be exported yet?

I try to modify the test.py script in order to export and I got to spill out the onnx model file, but it is kinda hard to understand the network input and output for a standalone implementation, I need this for TensorRT afterwards.  I noticed that the pytorch inference returns a tensor with 99 classes (Gym dataset) with probabilities and the result can be determined using a simple Argmax.

Any suggestions?

**Comment by kennymckormick** [comment]
1. It is just not implemented to be exported yet. 
2. The input of PoseC3D is a 5D volume: Batch x K (#keypoints) x T (temporal length) x H x W, the output is Batch x C (#classes)

**Comment by dragen1860** [comment]
@vsantosu You do not need to use mmaction2's onnx support. Just torch.onnx.export(raw_model). It's a simple conv3d model, without any difference with other models. you need to set the dummy input to the correct format.

**Comment by vsantosu** [comment]
@dragen1860 thanks for the tip! I got to export the ONNX model yesterday and I reversed engineered some of the additional post processing steps using the mmaction2 codebase.

If using the ONNX model, after we get results form inference the shape will be:

`
[batch_size, 20, 99] <- for gym model
`
This resulting tensor needs to be passed through a softmax function on the third axis, then apply an average on the second axis in order to get the corresponding predicted class using something like the ArgMax function to get the index of the highest probability.

I still have one doubt tho... The input test data has the following shape:

`
[1, 20, 17, 48, 64, 64]
` 
Dimension 0 is the batch size, dimension 1 is the temporal length like @kennymckormick said, in this case 20... Dimension 2 is the 17 keypoints. The last 2 dimensions are the height and width of the heatmaps for the heatmaps. What is exactly is dimension 3 (48)... Is this something I can obtain via the inference of HRNet or something? I'm guessing some types of channels but I'm not sure.

**Comment by dragen1860** [comment]
> @dragen1860 thanks for the tip! I got to export the ONNX model yesterday and I reversed engineered some of the additional post processing steps using the mmaction2 codebase.
> 
> If using the ONNX model, after we get results form inference the shape will be:
> 
> `[batch_size, 20, 99] <- for gym model`
> This resulting tensor needs to be passed through a softmax function on the third axis, then apply an average on the second axis in order to get the corresponding predicted class using something like the ArgMax function to get the index of the highest probability.
> 
> I still have one doubt tho... The input test data has the following shape:
> 
> `[1, 20, 17, 48, 64, 64]`
> Dimension 0 is the batch size, dimension 1 is the temporal length like @kennymckormick said, in this case 20... Dimension 2 is the 17 keypoints. The last 2 dimensions are the height and width of the heatmaps for the heatmaps. What is exactly is dimension 3 (48)... Is this something I can obtain via the inference of HRNet or something? I'm guessing some types of channels but I'm not sure.

Yes，dim 3 = 48 is a little bit different. You may notice there is 2 config parameters in config file: clips=10, double=True, These two parameter means for each single input sequences, it will sample 10 times, each time get 48 frames. and for each sampling, it will double again. so the total sequenches num is 10*2, and each sequences' frames len is 48.

**Comment by qijiahui7** [comment]
> @vsantosu You do not need to use mmaction2's onnx support. Just torch.onnx.export(raw_model). It's a simple conv3d model, without any difference with other models. you need to set the dummy input to the correct format.

How to export ONNX? I tried the usual method but it showed that the model does not exist. I loaded the config of poseC3D and printed the model, but it returned "none".


cfg = mmcv.Config.fromfile('/home/songkk/mmaction2-main/configs/skeleton/posec3d/slowonly_r50_8xb16-u48-240e_ntu60-xsub-keypoint.py')


model = build_model(cfg.model)
print(model)  #None
model.eval()
